File: backend/apps/finance/routers/analytics.py
from decimal import Decimal
from typing import List, Optional, Union

from ninja import Router, Schema

from finance.services import analytics_service

router = Router(tags=["finance:analytics"])


class KpiSerie(Schema):
    valor: Union[int, Decimal]
    formato: str  # "inteiro" | "moeda"
    serie: List[Union[int, Decimal]]
    labels: List[str]


class KpiOverviewSet(Schema):
    visitas: KpiSerie
    vendas: KpiSerie
    receita: KpiSerie
    ticket_medio: KpiSerie


class BehaviorItem(Schema):
    label: str
    valor: int


class ConversionItem(Schema):
    label: str
    valor_percentual: float
    detalhe: Optional[str] = None


class AnalyticsOverviewResponse(Schema):
    ultima_atualizacao: str
    kpis: KpiOverviewSet
    comportamento_visitantes: List[BehaviorItem]
    comportamento_checkout: List[BehaviorItem]
    conversoes: List[ConversionItem]


class ProdutoRanking(Schema):
    sku: str
    nome: str
    marca: str
    valor_principal: str
    valor_secundario: Optional[str] = None


class AnalyticsProdutosResponse(Schema):
    mais_vendidos: List[ProdutoRanking]
    mais_visualizados: List[ProdutoRanking]
    estoque_critico: List[ProdutoRanking]
    margem_ranking: List[ProdutoRanking]


@router.get("/overview", response=AnalyticsOverviewResponse)
def overview(request):
    """Estatísticas estilo Nuvemshop — visitas, comportamento, checkout e conversões."""
    return analytics_service.build_overview()


@router.get("/products", response=AnalyticsProdutosResponse)
def products(request):
    """Rankings de produtos: mais vendidos, mais visualizados, estoque crítico e margem."""
    return analytics_service.build_products()

File: frontend/src/types/analytics.ts
export type KpiFormato = 'inteiro' | 'moeda'

export interface KpiSerie {
  valor: number | string
  formato: KpiFormato
  serie: Array<number | string>
  labels: string[]
}

export interface KpiOverviewSet {
  visitas: KpiSerie
  vendas: KpiSerie
  receita: KpiSerie
  ticket_medio: KpiSerie
}

export interface BehaviorItem {
  label: string
  valor: number
}

export interface ConversionItem {
  label: string
  valor_percentual: number
  detalhe?: string | null
}

export interface AnalyticsOverview {
  ultima_atualizacao: string
  kpis: KpiOverviewSet
  comportamento_visitantes: BehaviorItem[]
  comportamento_checkout: BehaviorItem[]
  conversoes: ConversionItem[]
}

export interface ProdutoRanking {
  sku: string
  nome: string
  marca: string
  valor_principal: string
  valor_secundario?: string | null
}

export interface AnalyticsProdutos {
  mais_vendidos: ProdutoRanking[]
  mais_visualizados: ProdutoRanking[]
  estoque_critico: ProdutoRanking[]
  margem_ranking: ProdutoRanking[]
}
